[rb] Add FedCM support to the ruby selenium client

aguspe opened the pull request:
### **User description**
### Description
As mentioned in #12088 

The goal is to add support for the Federal credential management API (FedCM) so users have access to commands that will enable them to automate it

Reference Docs: 

https://fedidcg.github.io/FedCM/#automation
https://developer.mozilla.org/en-US/docs/Web/API/FedCM_API


### Motivation and Context
On Chrome 108 FedCM has shipped and it's already implemented by several web solutions, by adding support to the client library we guarantee that more users can automate this functionality without work arounds

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
<!--- Provide a general summary of your changes in the Title above -->


___

### **PR Type**
Enhancement, Tests


___

### **Description**
- Added support for Federal Credential Management (FedCM) API to the Ruby Selenium client.
- Implemented `HasFedCmDialog` extension for handling FedCM dialogs.
- Added `Account` and `Dialog` classes to represent and handle FedCM dialogs.
- Updated bridge to include methods for interacting with FedCM dialogs.
- Added integration and unit tests for FedCM dialog interactions.
- Updated FedCM configuration and type signatures.



___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><details><summary>13 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>driver.rb</strong><dd><code>Add FedCM dialog extension to Chromium driver</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
rb/lib/selenium/webdriver/chromium/driver.rb

<li>Added <code>DriverExtensions::HasFedCmDialog</code> to the list of extensions.<br>


</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13796/files#diff-9ef08354de5aef5f0c364d19d03f60a26409cdc19ca41f88f61032de7276dde0">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>common.rb</strong><dd><code>Require FedCM dialog and account in common driver</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
rb/lib/selenium/webdriver/common.rb

<li>Required <code>has_fedcm_dialog</code> extension.<br> <li> Required <code>fedcm/account</code> and <code>fedcm/dialog</code>.<br>


</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13796/files#diff-5bf2b6715d2e2a368c57757a030c83091c6d52e4db8175f66df8b98b12ac0d7a">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>has_fedcm_dialog.rb</strong><dd><code>Implement FedCM dialog handling extension</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
rb/lib/selenium/webdriver/common/driver_extensions/has_fedcm_dialog.rb

<li>Added <code>HasFedCmDialog</code> module with methods to handle FedCM dialogs.<br>


</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13796/files#diff-141873b56b208dce7e103aecb50610603cd78b381b366ac8b0b488ad9af3d4dd">+55/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>fedcm.rb</strong><dd><code>Autoload FedCM account and dialog modules</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
rb/lib/selenium/webdriver/common/fedcm.rb

- Added autoload for `Account` and `Dialog` under `FedCM` module.



</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13796/files#diff-19a5a7354a57e36a7f93dc1521b59fab897062e3310d84863e1a827523e2384d">+27/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>account.rb</strong><dd><code>Add FedCM account representation class</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
rb/lib/selenium/webdriver/common/fedcm/account.rb

- Added `Account` class to represent FedCM accounts.



</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13796/files#diff-501f36ef5dfefc11d9ef697487145382626c951dab349bf9f55db7dad7738841">+50/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>dialog.rb</strong><dd><code>Add FedCM dialog handling class</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
rb/lib/selenium/webdriver/common/fedcm/dialog.rb

- Added `Dialog` class to handle FedCM dialogs.



</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13796/files#diff-8ebd15a5560d28738b51288056a72a72a744ef1fe90b7eed4252a46e99f25cd0">+74/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>bridge.rb</strong><dd><code>Add FedCM dialog interaction methods to bridge</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
rb/lib/selenium/webdriver/remote/bridge.rb

- Added methods to handle FedCM dialog interactions.



</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13796/files#diff-42b562229083afc6ac8cc70a735fc1b24863402aa1404f08f60322b24cdfefc3">+40/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>commands.rb</strong><dd><code>Add FedCM dialog commands to bridge</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
rb/lib/selenium/webdriver/remote/bridge/commands.rb

- Added FedCM dialog related commands.



</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13796/files#diff-868e9bcf94865d4143225f9d7e3867b75a709d5a642a8fa1941e8625e54fad76">+13/-1</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>fedcm.html</strong><dd><code>Update FedCM config URL to use HTTP</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
common/src/web/fedcm/fedcm.html

- Updated `configURL` to use `http` instead of `https`.



</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13796/files#diff-8f9d11acba7737c2bddb11e8ae92668143c5549f56870e5f442ac92ce1052141">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>fedcm.json</strong><dd><code>Add login URL to FedCM configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
common/src/web/fedcm/fedcm.json

- Added `login_url` to FedCM configuration.



</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13796/files#diff-a01d9d9718317c5012496046288cc44a5d4841657663e56dc013b74d80142125">+2/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>account.rbs</strong><dd><code>Add type signatures for FedCM account class</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
rb/sig/lib/selenium/webdriver/fedcm/account.rbs

- Added type signatures for `Account` class.



</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13796/files#diff-9d232ead53e6a02cd48316241509846cb3811a308be9dcf23ff5e8c5b6eb1df0">+52/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>dialog.rbs</strong><dd><code>Add type signatures for FedCM dialog class</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
rb/sig/lib/selenium/webdriver/fedcm/dialog.rbs

- Added type signatures for `Dialog` class.



</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13796/files#diff-d2940b21ea43e490382b50f7e539f14a722097d455ecdb61ce20c14b4fec1b51">+26/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>bridge.rbs</strong><dd><code>Add type signatures for FedCM methods in bridge</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
rb/sig/lib/selenium/webdriver/remote/bridge.rbs

- Added type signatures for FedCM dialog methods in bridge.



</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13796/files#diff-beea6e4d56cd705f023d4c2ae2582ba9fd27da7007a5d803d8e6e793f18b87b0">+18/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Tests
</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>fedcm_spec.rb</strong><dd><code>Add integration tests for FedCM dialog</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
rb/spec/integration/selenium/webdriver/fedcm_spec.rb

- Added integration tests for FedCM dialog interactions.



</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13796/files#diff-222b9689cd84a69ef74f863584e09a6acf34a9c6a237610a4bd851dde791a22a">+90/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>account_spec.rb</strong><dd><code>Add unit tests for FedCM account class</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
rb/spec/unit/selenium/webdriver/common/fedcm/account_spec.rb

- Added unit tests for `Account` class.



</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13796/files#diff-b368aef668063d46f3d72a18dd04e30b43835aebd4881066401fce087ad7ff95">+54/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>dialog_spec.rb</strong><dd><code>Add unit tests for FedCM dialog class</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
rb/spec/unit/selenium/webdriver/common/fedcm/dialog_spec.rb

- Added unit tests for `Dialog` class.



</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13796/files#diff-bed3697a21ec1c4f606362c2e731712c37c8dc011d7b7d4fbe286b534538ded8">+93/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

p0deje commented on rb/lib/selenium/webdriver/fedcm.rb:
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
```
Can you move this file to `common` folder?

p0deje commented on rb/lib/selenium/webdriver/common/driver_extensions/has_fedcm_dialog.rb:
```diff
@@ -0,0 +1,36 @@
+        end
+
+        def fedcm_dialog
+          @fedcm_dialog ||= Selenium::WebDriver::FedCM::Dialog.new(@bridge)
```
You can avoid `Selenium::WebDriver::` here since it's the code is already within this module.

p0deje commented on rb/lib/selenium/webdriver/common/driver_extensions/has_fedcm_dialog.rb:
```diff
@@ -0,0 +1,36 @@
+        end
+
+        def wait_for_fedcm_dialog(timeout: 5, interval: 0.2, message: nil, ignore: nil)
+          wait = Selenium::WebDriver::Wait.new(timeout: timeout, interval: interval, message: message, ignore: ignore)
```
You can avoid Selenium::WebDriver:: here since it's the code is already within this module.

p0deje commented on rb/lib/selenium/webdriver/fedcm/account.rb:
```diff
@@ -0,0 +1,31 @@
+module Selenium
```
Can you move this directory to `common/fedcm` folder?

p0deje commented on rb/sig/selenium/web_driver/script.rbs:
```diff
@@ -1,20 +0,0 @@
-module Selenium
```
I think you removed it by accident, can you pleas restore?

aguspe replied:
Ups my bad, totally an accident

p0deje commented on rb/spec/integration/selenium/webdriver/fedcm/dialog_spec.rb:
```diff
@@ -0,0 +1,98 @@
+  module WebDriver
+    module FedCM
+      describe Dialog, exclusive: {browser: :chrome} do
+        let(:url) { 'https://fedcm-rp-demo.glitch.me/' }
```
Unfortunately, our tests should be atomic and not depend on external websites. Is it possible to replicate a demo similar to how we test basic authentication (https://github.com/SeleniumHQ/selenium/blob/29949525e37b36c9954285cfc500615c760d500e/rb/spec/integration/selenium/webdriver/spec_support/rack_server.rb#L122-L123) - purely locally without external network calls?

aguspe replied:
That makes a lot of sense, I will give this a look tomorrow to this, to see if I can replicate the flow without any external calls

p0deje commented on rb/spec/unit/selenium/webdriver/fedcm/account_spec.rb:
```diff
@@ -0,0 +1,45 @@
+          expect(account.privacy_policy_url).to eq('privacy-policy-url')
+        end
+
+        it 'has a constant LOGIN_STATE_SIGNIN' do
```
I don't think it is important to test constant values, we are just testing implementation detail.

p0deje commented on rb/spec/unit/selenium/webdriver/fedcm/dialog_spec.rb:
```diff
@@ -0,0 +1,74 @@
+  module WebDriver
+    module FedCM
+      describe Dialog do
+        let(:bridge) { instance_double(Selenium::WebDriver::Remote::Bridge) }
```
You can avoid `Selenium::WebDriver::` here since it's the code is already within this module.

p0deje commented on rb/spec/integration/selenium/webdriver/fedcm/dialog_spec.rb:
```diff
@@ -0,0 +1,98 @@
+module Selenium
+  module WebDriver
+    module FedCM
+      describe Dialog, exclusive: {browser: :chrome} do
```
Let's switch from `exclusive` to `only` - this will ensure tests are run in other browsers but expected to fail.

aguspe commented:
> This definitely looks like a great start, thank you for working on this!
> 
> Can we add more tests to verify all the endpoints such as delays and cooldowns?

I will add the extra tests tomorrow morning, I added a thumbs up to all the resolved comments, so anything else I can improve or fix just let me know

Thank you so much @p0deje for the quick review! super appreciated, and thanks for taking the time to help me out :)

aguspe commented:
@p0deje I tweaked a bit the json files and the HTML  from this PR  #12096  , which was a great inspiration, because if not I couldn't get the dialog to pop up and there was an endpoint missing that crashes the function triggerFedCm()

Also, I added the extra tests you asked, but I'm having the challenge that every endpoint that requires for me to pass the body such as https://fedidcg.github.io/FedCM/#webdriver-setdelayenabled

This throws me this error from this method: 

`

 ```
 ArgumentError: {:enabled=>true} invalid for :set_fedcm_delay
 
 0) Selenium::WebDriver::FedCM with dialog present sets the delay
     Failure/Error: raise ArgumentError, "#{opts.inspect} invalid for #{command.inspect}"

     ArgumentError:`
```

The source of the error is this method:
     
     `       def execute(command, opts = {}, command_hash = nil)
          verb, path = commands(command) || raise(ArgumentError, "unknown command: #{command.inspect}")
          path = path.dup

          path[':session_id'] = session_id if path.include?(':session_id')

          begin
            opts.each { |key, value| path[key.inspect] = escaper.escape(value.to_s) }
          rescue IndexError
            raise ArgumentError, "#{opts.inspect} invalid for #{command.inspect}"
          end

          WebDriver.logger.debug("-> #{verb.to_s.upcase} #{path}", id: :command)
          http.call(verb, path, command_hash)['value']
        end
       
  I will keep working on this tomorrow, but if you could give me a clue it would be greatly appreciated haha, the rest of the tests are passing though

p0deje commented:
> I will keep working on this tomorrow, but if you could give me a clue it would be greatly appreciated haha, the rest of the tests are passing though

It's hard to say, but maybe you need to wrap `enabled: enabled` with `{}` to make it be treated as a hash.

aguspe commented:
> > I will keep working on this tomorrow, but if you could give me a clue it would be greatly appreciated haha, the rest of the tests are passing though
> 
> It's hard to say, but maybe you need to wrap `enabled: enabled` with `{}` to make it be treated as a hash.

@p0deje the issue was that I needed to pass the parameters as a command hash and not as options

This works:

`execute :select_fedcm_account, {},{ accountIndex: index }`

Thanks for the help and it's ready for a re-review :)

aguspe commented:
Hi everyone, I hope you had a nice weekend

This PR is up for re-review @p0deje @titusfortner :)

I will keep reading the FedCM docs regarding click button because it seems to be an issue with the mock but all the other tests pass when executed

p0deje commented on rb/spec/integration/selenium/webdriver/fedcm_spec.rb:
```diff
@@ -0,0 +1,86 @@
+        end
+
+        context 'without dialog present' do
+          it 'throws an error when getting the title' do
```
Because the integration tests are expensive and require execution of all hooks/restart a browser in case of failure, we prefer to consolidate tests whenever possible. For example, in this case, all raise error tests can be joined into a single test:

```ruby
context 'without dialog present' do
  it 'throws an error' do
    expect { dialog.title }.to raise_error(Error::NoSuchAlertError)
    expect { dialog.subtitle }.to raise_error(Error::NoSuchAlertError)
    ...
```

The rest of the integration can be joined as well around their behavior.

p0deje commented on rb/lib/selenium/webdriver/common/driver_extensions/has_fedcm_dialog.rb:
```diff
@@ -0,0 +1,36 @@
+module Selenium
```
Please add the following to the top of every new file you've added:

```ruby
# frozen_string_literal: true

# Licensed to the Software Freedom Conservancy (SFC) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The SFC licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
```

p0deje commented on rb/lib/selenium/webdriver/common/fedcm/account.rb:
```diff
@@ -0,0 +1,31 @@
+      # See: https://fedidcg.github.io/FedCM/#dictdef-identityprovideraccount
+      #      https://fedidcg.github.io/FedCM/#webdriver-accountlist
+      class Account
+        LOGIN_STATE_SIGNIN = 'SignIn'.freeze
```
Rather than freezing strings manually, you can add the directive to the top:

```ruby
# frozen_string_literal: true
```

p0deje commented on rb/lib/selenium/webdriver/common/fedcm/dialog.rb:
```diff
@@ -0,0 +1,57 @@
+
+        # Selects an account as if the user had clicked on it.
+        #
+        # @param [Integer] index The index of the account to select from the list returned by get_accounts.
```
Should this accept `Account` as a parameter instead? This would make the API more user-friendly:


```ruby
accounts = driver.fedcm_dialog.accounts
driver.fedcm_dialog.select_accounts(accounts.first)
```

aguspe replied:
The thing is that when checking how is described on the docs https://fedidcg.github.io/FedCM/#webdriver-selectaccount:

```MARKDOWN
The [remote end steps](https://w3c.github.io/webdriver/#dfn-remote-end-steps) are:

If parameters is not a JSON [Object](https://tc39.es/ecma262/multipage/structured-data.html#sec-json-object), return a [WebDriver error](https://w3c.github.io/webdriver/#dfn-error) with [error code](https://w3c.github.io/webdriver/#dfn-error-code) [invalid argument](https://w3c.github.io/webdriver/#dfn-invalid-argument).

If no FedCM dialog is currently open, return a [WebDriver error](https://w3c.github.io/webdriver/#dfn-error) with [error code](https://w3c.github.io/webdriver/#dfn-error-code) [no such alert](https://w3c.github.io/webdriver/#dfn-no-such-alert).

Let accountIndex be the result of [getting a property](https://w3c.github.io/webdriver/#dfn-getting-properties) named "accountIndex" from parameters.

If accountIndex is [undefined](https://webidl.spec.whatwg.org/#idl-undefined) or is less than 0 or greater than or equal to the number of accounts that the user can choose from in the current flow, return a [WebDriver error](https://w3c.github.io/webdriver/#dfn-error) with [error code](https://w3c.github.io/webdriver/#dfn-error-code) [invalid argument](https://w3c.github.io/webdriver/#dfn-invalid-argument).

Close the dialog and continue the [create an IdentityCredential](https://fedidcg.github.io/FedCM/#create-an-identitycredential) algorithm as if the user had selected the account indicated by accountIndex and [granted permission to sign-up](https://fedidcg.github.io/FedCM/#request-permission-to-sign-up), if applicable.

Return [success](https://w3c.github.io/webdriver/#dfn-success) with data null.
```


So if we would like to provide the user the opportunity to pass an account our method should look like this:

```RUBY
  def select_account(account)
    account_index = accounts.index(account)
    @bridge.select_fedcm_account account_index
  end
  ```
  
  Is that something we want to do? because then we will also need error handling to clarify that the account is not present

p0deje replied:
Ok, let's keep using `index` instead - that's also what Java does, we can provide convenient API later.

p0deje commented:
Can you please take care of the CI failures (Ruby-related)?

aguspe commented:
> Can you please take care of the CI failures (Ruby-related)?

Yes for sure, I will start fixing them now and commit tomorrow :) thank you for the help

titusfortner commented:
Thanks for your continued work on this; please remember to run rubocop as you fix issues.

aguspe commented:
> Thanks for your continued work on this; please remember to run rubocop as you fix issues.

Thank you for the help and that's a great point, sorry that I forgot to run rubocop, and @p0deje I noticed that edge still tried to execute the fedcm integration tests even though is only for chrome so I adjusted that and I moved the integration tests to the chrome folder

p0deje commented:
> I noticed that edge still tried to execute the fedcm integration tests even though is only for chrome so I adjusted that and I moved the integration tests to the chrome folder

Actually, I think what you were seeing is that your implementation _actually works in Edge_, while you marked it to work only in Chrome. You should move the tests back from Chrome folder and mark it as `only: {browser: %w[chrome edge]}`. The error in CI was suggesting that:

```
expected pending 'Test guarded; Guarded by {:browser=>:chrome, :reason=>"No reason given"};' to fail. No error was raised.
```

aguspe commented:
> {browser: %w[chrome edge]}



> > I noticed that edge still tried to execute the fedcm integration tests even though is only for chrome so I adjusted that and I moved the integration tests to the chrome folder
> 
> Actually, I think what you were seeing is that your implementation _actually works in Edge_, while you marked it to work only in Chrome. You should move the tests back from Chrome folder and mark it as `only: {browser: %w[chrome edge]}`. The error in CI was suggesting that:
> 
> ```
> expected pending 'Test guarded; Guarded by {:browser=>:chrome, :reason=>"No reason given"};' to fail. No error was raised.
> ```

Thank you so much for the explanation, what throw me off before is that I tried this, but now that I looked more into the Guard implementation, this works and the test pass both with edge and chrome:

`describe FedCM, exclusive: [{bidi: false, reason: 'Not yet implemented with BiDi'}, {browser: %i[chrome edge]}] `

aguspe commented:
Is there anything I can do to help this get merged? All the ruby tests seemed to have passed in the last run, @p0deje @titusfortner

p0deje commented:
Thank you for the contribution!
